### ISSUE TITLE
BUG: S3 watcher silently failing if bucket name contains periods (.)

### ISSUE THREAD

**Issue by kirk-snapper** [Netflix/security_monkey#430]
Hi,

In the s3 watcher:
https://github.com/Netflix/security_monkey/blob/develop/security_monkey/watchers/s3.py#L141-L144

There's some code that looks up the region for a bucket:

loc = self.wrap_aws_rate_limited_call(bucket.get_location)
region = self.translate_location_to_region(loc)

This appears to fail if the bucket name contains dots/periods/.'s, as the certificate validation fails.

I have pasted the message and stacktrace that are logged in the exceptions table below.

It appears to be a known issue in the boto library that Security Monkey uses:  https://github.com/boto/boto/issues/2836#issuecomment-68608362

I am currently working around it by disabling certificate validation per the comment linked to above.

The behaviour observed within Security Monkey is that the S3 bucket shows as "Deleted", and so in my case a bucket with audit issues didn't show in the results. Perhaps if an exception is caught and logged, there should be something visible in the UI to alert the user?

Thanks,

Kirk

Message and stacktrace:

```
'AWS returned an exception while attempting to obtain information on a bucket I should have access to. Bucket Name: bucket.name.example'

Traceback (most recent call last):                                                                                                         +
   File "/usr/local/src/security_monkey-develop/security_monkey/watchers/s3.py", line 142, in slurp                                         +
     loc = self.wrap_aws_rate_limited_call(bucket.get_location)                                                                             +
   File "/usr/local/src/security_monkey-develop/security_monkey/watcher.py", line 112, in wrap_aws_rate_limited_call                        +
     retval = awsfunc(*args, **nargs)                                                                                                       +
   File "/usr/local/lib/python2.7/dist-packages/boto-2.41.0-py2.7.egg/boto/s3/bucket.py", line 1134, in get_location                        +
     query_args='location')                                                                                                                 +
   File "/usr/local/lib/python2.7/dist-packages/boto-2.41.0-py2.7.egg/boto/s3/connection.py", line 668, in make_request                     +
     retry_handler=retry_handler                                                                                                            +
   File "/usr/local/lib/python2.7/dist-packages/boto-2.41.0-py2.7.egg/boto/connection.py", line 1071, in make_request                       +
     retry_handler=retry_handler)                                                                                                           +
   File "/usr/local/lib/python2.7/dist-packages/boto-2.41.0-py2.7.egg/boto/connection.py", line 943, in _mexe                               +
     request.body, request.headers)                                                                                                         +
   File "/usr/lib/python2.7/httplib.py", line 1057, in request                                                                              +
     self._send_request(method, url, body, headers)                                                                                         +
   File "/usr/lib/python2.7/httplib.py", line 1097, in _send_request                                                                        +
     self.endheaders(body)                                                                                                                  +
   File "/usr/lib/python2.7/httplib.py", line 1053, in endheaders                                                                           +
     self._send_output(message_body)                                                                                                        +
   File "/usr/lib/python2.7/httplib.py", line 897, in _send_output                                                                          +
     self.send(msg)                                                                                                                         +
   File "/usr/lib/python2.7/httplib.py", line 859, in send                                                                                  +
     self.connect()                                                                                                                         +
   File "/usr/lib/python2.7/httplib.py", line 1278, in connect                                                                              +
     server_hostname=server_hostname)                                                                                                       +
   File "/usr/lib/python2.7/ssl.py", line 353, in wrap_socket                                                                               +
     _context=self)                                                                                                                         +
   File "/usr/lib/python2.7/ssl.py", line 601, in __init__                                                                                  +
     self.do_handshake()                                                                                                                    +
   File "/usr/lib/python2.7/ssl.py", line 838, in do_handshake                                                                              +
     match_hostname(self.getpeercert(), self.server_hostname)                                                                               +
   File "/usr/lib/python2.7/ssl.py", line 272, in match_hostname                                                                            +
     % (hostname, ', '.join(map(repr, dnsnames))))                                                                                          +
 CertificateError: hostname u'bucket.name.example.s3.amazonaws.com' doesn't match either of '*.s3.amazonaws.com', 's3.amazonaws.com'+
```

**Comment by mikegrima** [comment]
There are a couple of exceptions that are raised by the boto2 S3 library. We should really focus on moving that to boto3.

**Comment by monkeysecurity** [comment]
yikes.  

So we do have buckets with periods in their names, and security_monkey is handling them correctly as far as I can tell.  This is one of the oldest watchers in security_monkey, and it would be really nice to move it to boto3.

Could you provide me with a `pip freeze | grep boto`

**Comment by kirk-snapper** [comment]
Thanks Patrick,

I wonder if it's to do with the version of python? The boto ticket suggests that cert checking might have been strengthened in 2.7.9.

$ pip freeze | grep boto
You are using pip version 8.1.1, however version 8.1.2 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
boto==2.41.0
boto3==1.3.1
botocore==1.4.37
botor==0.0.1.dev8

$ python --version
Python 2.7.12

Cheers,

Kirk

**Comment by monkeysecurity** [comment]
I'm just barely ahead:

```
boto==2.42.0
boto3==1.4.0
botocore==1.4.56
botor==0.0.1.dev8
```

I'll label this as a bug and we'll start rebuilding the s3 watcher in boto3.  @mikegrima already has something similar done for a separate project.

**Comment by aebie** [comment]
This is showing up ad an issue in our deployment as well. I have a task set up to look into it so will let you know what we find.

**Comment by monkeysecurity** [comment]
This was fixed in PR #443 which was merged to develop 16 days ago.

We'll have a release out sometime in December.

Closing ticket.
